src/EditorFeatures/VisualBasic/LineCommit/CommitBufferManager.vb

```diff
@@ -97,63 +96,119 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.LineCommit
-                    Dim info As FormattingInfo
-                    If Not TryComputeExpandedSpanToFormat(dirtyRegion, info, cancellationToken) Then
-                        Return
-                    End If
-
-                    Dim useSemantics = info.UseSemantics
-                    If useSemantics AndAlso Not isExplicitFormat Then
-                        ' Avoid using semantics for formatting extremely large dirty spans without an explicit request
-                        ' from the user. The "large span threshold" is 7000 lines. The 7000 line threshold is an
-                        ' estimated value accounting for a lower-bound of the algorithmic complexity of text
-                        ' differencing in designer cases along with measurements of a pathological example demonstrated
-                        ' at 14000 lines. We expect Windows Forms designer formatting operations to run in under ~15
-                        ' seconds on average current hardware when nearing the threshold.
-                        Dim startLineNumber = 0
-                        Dim startCharIndex = 0
-                        Dim endLineNumber = 0
-                        Dim endCharIndex = 0
-                        info.SpanToFormat.GetLinesAndCharacters(startLineNumber, startCharIndex, endLineNumber, endCharIndex)
-                        If endLineNumber - startLineNumber > 7000 Then
-                            useSemantics = False
-                        End If
-                    End If
+            Return CommitDirtyWorkerAsync(isExplicitFormat, blocking, cancellationToken)
+        End Function
 
-                    Dim tree = _dirtyState.BaseDocument.GetSyntaxTreeSynchronously(cancellationToken)
-                    _commitFormatter.CommitRegion(info.SpanToFormat, isExplicitFormat, useSemantics, dirtyRegion, _dirtyState.BaseSnapshot, tree, cancellationToken)
+        Private Async Function CommitDirtyWorkerAsync(
+                isExplicitFormat As Boolean,
+                blocking As Boolean,
+                cancellationToken As CancellationToken) As Task
+            Me.AssertIsForeground()
+
+            ' Create our own linked token to do the async committing work.  This way we can cancel the work if we see
+            ' any edits come through to the buffer.
+
+            Using bufferEditTokenSource = New CancellationTokenSource()
+                Using linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(bufferEditTokenSource.Token, cancellationToken)
+
+                    ' if the text buffer is changed, then cancel any work we're doing.
+                    Dim textBuffer = _dirtyState.BaseSnapshot.TextBuffer
+                    Dim handler = Sub(sender As Object, eventArgs As TextContentChangingEventArgs)
+                                      bufferEditTokenSource.Cancel()
+                                  End Sub
+                    AddHandler textBuffer.Changing, handler
```

Comment:
we are not limited to listening to just this event.  However, it seemed the most basic and most reasonable indicator that the user made a change and we should stop our work.  We could also just literally ignore all events and let all the work proceed (since it will just fail when trying to apply later).  However, that seems somewhat wasteful, especially as this work may get semantics and we would not want that to just be consuming BG resources indefinitely.